src/linear_algebra/cayley_hamilton.lean

```diff
@@ -0,0 +1,96 @@
+/-
+Copyright (c) 2020 Scott Morrison. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Scott Morrison
+-/
+import tactic.apply_fun
+import ring_theory.matrix_algebra
+import ring_theory.polynomial_algebra
+import linear_algebra.nonsingular_inverse
+
+/-!
+# The Cayley-Hamilton theorem, over a commutative ring.
+
+We use a nice proof from http://drorbn.net/AcademicPensieve/2015-12/CayleyHamilton.pdf
+-/
+
+noncomputable theory
+
+universes u v w
+
+open polynomial matrix
+open_locale big_operators
+
+variables {R : Type u} [comm_ring R]
+variables {n : Type w} [fintype n] [decidable_eq n]
+
+open finset
+
+/--
+The "characteristic matrix" of `M : matrix n n R` is the matrix of polynomials $t I - M$.
+The determinant of this matrix is the characteristic polynomial.
+-/
+def char_matrix (M : matrix n n R) : matrix n n (polynomial R) :=
+matrix.scalar n (X : polynomial R) - (λ i j, C (M i j))
```

Comment:
Is it worth defining `matrix.map : (R →+* S) → matrix n n R →+* matrix n n S` and using this to define this function?

Reply:
`matrix.map` should definitely exist.

Reply:
Okay, I've done that as #3352, and will merge that into this, and refactor.

Reply:
Okay, done.